### DISCUSSION TITLE
Export as loop (remove end silence)

### DISCUSSION THREAD

**Discussion by j36e** [FluidSynth/fluidsynth#1313] [Q&A]
Hello

I have been trying to render midi files to audio using fluidsynth with commands such as `fluidsynth.exe -T wav -F output.wav soundfount.sf2 midi.mid`. It works fine, except for the ~7 seconds silence at the end that I would like to get read of. I know I could just cut it out in another program but I need the song to end exactly at the end so that it could loop.

I think I looked everywhere the wiki, discussions, man page, stackoverflow, etc... and couldn't find anything helpful. I know this is feasable, vlc has a flush method in their source code that I think does what I need but I can't understand how it works lmao (LMMS, a program to make music, also has an option to cut the silence when exporting midi to wav).

If anyone knows how to export a midi file without the trailing silence I would greatly appreciate the help.

Thanks a lot for reading me

**Comment by mawe42** [comment]
I think it depends on what exactly you want to achieve. You say you want the wav file to loop, which would imply that you need the wav file to end musically at the end of the last beat. If that is your goal, then trimming the silence at the end of the exported wav will probably not (always) be the right solution. If the last sample in a 4/4 song plays on beat 3 and is quite short, then you would cut the song too early. If the last sample plays on beat four and has a long reverb phase, then the resulting wav would be too long. In both cases, you would have a noticable "jump" in your loop. Looping the MIDI file might be the easier way forward here.

But if your goal is really just to remove silence at the end of the wav, you could simply use sox to remove it automatically. sox is also available for windows. This guide might be helpful: https://digitalcardboard.com/blog/2009/08/25/the-sox-of-silence/comment-page-2/

**Reply by j36e** [reply]
Hi and thank you for your answer but I do need the audio to end at the exact right timing regardless of reverb and stuff so your solution won't do sorry :(

I just need fluidsynth to stop rendering when there is nothing to be rendered

**Comment by derselbst** [comment]
The MIDI player found in recent versions of fluidsynth renders as long as there are voices playing (i.e. producing sound). Potential MIDI CCs at the very end can also cause extending the playtime. Without additional information, like the original MIDI file, the version of fluidsynth you're using, etc. it's not possible to tell what's going on.

Yet, since you're up for looping, simply relying on a proper ending is error prone, esp when having reverb and chorus enabled, like you do. Instead, you should ensure that your MIDI file contains proper markers, that some application (potentially one you have written) correctly interprets as loop markers and acts accordingly. E.g. by exporting the audio as wave file with proper smpl-chunk in place.

**Comment by SimonBelmontis** [comment]
I am currently having the same issue. Trying to convert midi files of the original Final Fantasy 7 with the generic Midi Synthesizer of Windows and each one of them have a trailing end of silence.

Example: 
[TrailingSilenceExample.zip](https://github.com/FluidSynth/fluidsynth/files/15203007/TrailingSilenceExample.zip)

Midi: GM - 1-10 Fighting.mid
SoundFond: gm.sf2
Command: fluidsynth -nli -r 48000 -o synth.cpu-cores=2 -T oga -F "bat.ogg" "gm.sf2" "GM - 1-10 Fighitng.mid"
